components/datetime/src/pattern/parser.rs

```diff
@@ -0,0 +1,118 @@
+use super::error::Error;
+use super::{Pattern, PatternItem};
+use crate::fields::{Field, FieldLength, FieldSymbol};
+use std::convert::TryFrom;
+
+enum State {
+    Symbol {
+        symbol: FieldSymbol,
+        byte: u8,
+        start_idx: usize,
+    },
+    Literal {
+        start_idx: usize,
+    },
+}
+
+pub struct Parser<'p> {
+    source: &'p [u8],
+    state: State,
+    items: Vec<PatternItem>,
+    ptr: usize,
+}
+
+impl<'p> Parser<'p> {
+    pub fn new(source: &'p [u8]) -> Self {
+        Self {
+            source,
+            state: State::Literal { start_idx: 0 },
+            items: Vec::with_capacity(source.len()),
+            ptr: 0,
+        }
+    }
+
+    fn collect_item(&mut self) -> Result<(), Error> {
+        match self.state {
+            State::Symbol {
+                symbol, start_idx, ..
+            } => self.collect_symbol(symbol, start_idx)?,
+            State::Literal { start_idx } => self.collect_literal(start_idx),
+        }
+        Ok(())
+    }
+
+    fn collect_symbol(&mut self, symbol: FieldSymbol, start_idx: usize) -> Result<(), Error> {
+        let len = self.ptr - start_idx;
+        let length = FieldLength::try_from(len)?;
+        self.items.push(Field { symbol, length }.into());
+        Ok(())
+    }
+
+    fn collect_literal(&mut self, start_idx: usize) {
+        if start_idx < self.ptr {
+            let slice = &self.source[start_idx..self.ptr];
+            let item =
+                PatternItem::Literal(unsafe { String::from_utf8_unchecked(slice.to_owned()) });
```

Comment:
This is unsafe.  The string comes from `pub fn from_bytes(input: &[u8])`, and it is not necessarily UTF-8.

Why don't you just use `&str` throughout?  What does `&[u8]` gain you?

Reply:
Char operations are slower by 5-10% because of indices. Since we're hunting ASCII bytes it is faster to iterate over bytes than guard against UTF8 indices.

I understand that there are UTF8 literals between symbols, but their boundaries should be pretty clear as they can only start after symbol and end on the next.

I crafted a replacement operating on char indices and I see ~8-10% perf hit.
Here it is: https://paste.mozilla.org/zEFKZ9rj

If you think this is an okay cost to pay, I'm okay putting it here. If not, I'm ok replacing `unchecked` with `lossy` (as it was before) which checks the literals (that costs 2-3%).

Lmk what you prefer.

Reply:
I'm not necessarily suggesting that you iterate over chars.  You can iterate over bytes in the string (using std::str::Bytes) and still retain type system safety.

Reply:
> I understand that there are UTF8 literals between symbols, but their boundaries should be pretty clear as they can only start after symbol and end on the next.

This is correct.  However, your code doesn't do anything to protect against, e.g., Latin1 input, which has the same ASCII bytes for the pattern symbols but non-UTF8 bytes for the literals.  Your code currently takes in a u8 buffer and returns a string without validating it.

Reply:
> I crafted a replacement operating on char indices and I see ~8-10% perf hit.

Do your benchmarks include data that includes non-ASCII characters?  Like the datetime patterns used for Chinese, French, and others?

Reply:
switched to `bytes()` as discussed on slack.